### PR TITLE
feat: adding Enum

### PR DESCRIPTION
**Pull request by pandadefi**
### What I did

Adding enums to vyper

fixes: #2319

### Commit message

Commit message for the final, squashed PR. (Optional, but reviewers will appreciate it! Please see [our commit message style guide](../../blob/master/docs/style-guide.rst#best-practices-1) for what we would ideally like to see in a commit message.)

### Cute Animal Picture

image

### CODE REVIEW

File: vyper/semantics/types/user/enum.py [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+        return Enum(abi["name"], abi["inputs"])
+
+    @classmethod
+    def from_EnumDef(cls, base_node: vy_ast.EventDef) -> "Enum":
+        """
+        Generate an `Event` object from a Vyper ast node.
+
+        Arguments
+        ---------
+        base_node : EventDef
```
Comment (fubuloubu) [inline vyper/semantics/types/user/enum.py:66]:
```suggestion
        base_node : EnumDef
```

**Comment by charles-cooper** [comment]
whoops, re-requested review but maybe needs some more tests. still in draft mode

**Comment by charles-cooper** [comment]
lgtm, but i contributed some work to this and so we should get a review from @fubuloubu as well.

**Review by charles-cooper — APPROVED** [review]

**Comment by charles-cooper** [comment]
one q i have is whether we should introduce any conventions around naming of enum members. options:
```
enum Action:
    buy
    sell
```
```
enum Action:
    Buy
    Sell
```
```
enum Action:
    BUY
    SELL
```

**Comment by fubuloubu** [comment]
> one q i have is whether we should introduce any conventions around naming of enum members.

Typically this is the way enums are usually specified

```vyper
enum Action:
    BUY
    SELL
```

File: docs/types.rst [hunk 2]
```diff
@@ -317,6 +317,27 @@ On the ABI level the Fixed-size bytes array is annotated as ``string``.
 
     example_str: String[100] = "Test String"
 
+Enums
+-----
+
+**Keyword:** ``enum``
+
+Enums are custom defined types. An enum must have at least one member, and can hold up to a maximum of 256 members.
+The members are represented by ``uint256`` values in the form of 2\ :sup:`N` where ``N`` is the index of the member (``1 <= N <= 256``).
```
Comment (fubuloubu) [inline docs/types.rst:326]: Typically when specifying concrete variables, you'd use lowercase names
```suggestion
The members are represented by ``uint256`` values in the form of 2\ :sup:`n` where ``n`` is the index of the member in the range ``0 <= n <= 255``.
```

File: docs/types.rst [hunk 3]
```diff
@@ -317,6 +317,27 @@ On the ABI level the Fixed-size bytes array is annotated as ``string``.
...
+    # Defining an enum with two members
+    enum Roles:
+        admin
+        user
+
+    # Declaring an enum variable
+    role: Roles = Roles.admin
+
+    # Returning a member
+    return Roles.admin
```
Comment (fubuloubu) [inline docs/types.rst:339]: To match with the previous comment about naming conventions
```suggestion
        ADMIN
        USER

    # Declaring an enum variable
    role: Roles = Roles.ADMIN

    # Returning a member
    return Roles.ADMIN
```

**Review by fubuloubu — APPROVED** [review]
Some nits on naming

**Pull request merged**